File: swift/Core/Sources/PrexpCore/Sort.swift
/// Process sort fields — cycled with `s` (and direction toggled with `S`),
/// mirroring the Rust front-ends. Numeric fields default to descending (biggest
/// first); name/pid default to ascending.
public enum SortField: CaseIterable, Sendable {
    case cpu, memory, pmem, threads, files, total, name, pid

    public var label: String {
        switch self {
        case .cpu: return "CPU%"
        case .memory: return "MEM"
        case .pmem: return "PMEM"
        case .threads: return "THR"
        case .files: return "FILES"
        case .total: return "TOTAL"
        case .name: return "NAME"
        case .pid: return "PID"
        }
    }

    /// Whether this field's natural order is descending.
    var descendingByDefault: Bool {
        switch self {
        case .name, .pid: return false
        default: return true
        }
    }

    public func next() -> SortField {
        let all = SortField.allCases
        let i = all.firstIndex(of: self)!
        return all[(i + 1) % all.count]
    }
}

/// Sort a snapshot list. `cpu` supplies the (tracker-derived) CPU% since it isn't
/// on the snapshot itself. Stable, and `reversed` flips the natural direction.
public func sortProcesses(_ procs: [ProcessSnapshot], by field: SortField,
                          reversed: Bool, cpu: (Int32) -> Double) -> [ProcessSnapshot] {
    let sorted: [ProcessSnapshot]
    switch field {
    case .cpu:
        sorted = procs.sorted { cpu($0.pid) > cpu($1.pid) }
    case .memory:
        sorted = procs.sorted { $0.memory.rss > $1.memory.rss }
    case .pmem:
        sorted = procs.sorted { $0.memory.phys > $1.memory.phys }
    case .threads:
        sorted = procs.sorted { $0.activity.threadCount > $1.activity.threadCount }
    case .files:
        sorted = procs.sorted { $0.count(of: .file) > $1.count(of: .file) }
    case .total:
        sorted = procs.sorted { $0.resources.count > $1.resources.count }
    case .name:
        sorted = procs.sorted { $0.name.lowercased() < $1.name.lowercased() }
    case .pid:
        sorted = procs.sorted { $0.pid < $1.pid }
    }
    // `descendingByDefault` fields are already in descending order above; reversed
    // flips whatever the natural order is.
    return reversed ? sorted.reversed() : sorted
}

/// Case-insensitive substring match on name or PID — the `/` search predicate.
public func matchesQuery(_ p: ProcessSnapshot, _ query: String) -> Bool {
    if query.isEmpty { return true }
    let q = query.lowercased()
    return p.name.lowercased().contains(q) || String(p.pid).contains(q)
}

File: rust/crates/prexp-desktop/src/app/sort.rs
use std::cmp::Reverse;
use std::fmt;

use prexp_core::models::{ProcessSnapshot, ResourceKind};

use super::cpu::CpuTracker;

/// A column the process table can be ordered by. Numeric fields default to
/// descending (biggest consumers first); `NAME`/`PID` default to ascending.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SortField {
    Cpu,
    Rss,
    Pmem,
    Threads,
    Files,
    Total,
    Name,
    Pid,
}

impl SortField {
    /// Every field, in dropdown order.
    pub const ALL: [SortField; 8] = [
        SortField::Cpu,
        SortField::Rss,
        SortField::Pmem,
        SortField::Threads,
        SortField::Files,
        SortField::Total,
        SortField::Name,
        SortField::Pid,
    ];

    fn label(self) -> &'static str {
        match self {
            SortField::Cpu => "CPU%",
            SortField::Rss => "MEM",
            SortField::Pmem => "PMEM",
            SortField::Threads => "THR",
            SortField::Files => "FILES",
            SortField::Total => "TOTAL",
            SortField::Name => "NAME",
            SortField::Pid => "PID",
        }
    }
}

impl fmt::Display for SortField {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.label())
    }
}

/// Order `procs` in place by `field`. Each field has a natural default
/// direction (numeric = descending, textual = ascending); `reverse` flips it.
pub fn sort_processes(
    procs: &mut [ProcessSnapshot],
    field: SortField,
    cpu: &CpuTracker,
    reverse: bool,
) {
    // `Reverse` keys give the numeric fields their descending default; CPU% is
    // an `f64`, so it uses `total_cmp` (there's no `Ord` key to reverse).
    match field {
        SortField::Name => procs.sort_by_key(|p| p.name.to_lowercase()),
        SortField::Pid => procs.sort_by_key(|p| p.pid),
        SortField::Cpu => {
            procs.sort_by(|a, b| cpu.percent(b.pid).total_cmp(&cpu.percent(a.pid)))
        }
        SortField::Rss => procs.sort_by_key(|p| Reverse(p.memory.rss)),
        SortField::Pmem => procs.sort_by_key(|p| Reverse(p.memory.phys)),
        SortField::Threads => procs.sort_by_key(|p| Reverse(p.activity.thread_count)),
        SortField::Files => procs.sort_by_key(|p| Reverse(p.count_by_kind(&ResourceKind::File))),
        SortField::Total => procs.sort_by_key(|p| Reverse(p.resources.len())),
    }
    if reverse {
        procs.reverse();
    }
}
